# Тема 10. Декораторы и исключения.
### Отчет по Теме #10 выполнил:
- Смотрин Матвей Станиславович
- ПИЭ-21-2

| Задание | Лаб. раб. | Сам. раб. |
| ------ | ------ | ------ |
| Задание 1 | + | + |
| Задание 2 | + | + |
| Задание 3 | + | + |
| Задание 4 | + | + |
| Задание 5 | + | + |

Работу проверили:
- к.э.н., доцент Панов М.А.

## Лабораторная работа 
## Задание №1
Наверняка вы думаете, что декораторы – это какая-то бесполезная
вещь, которая вам никогда не пригодится, но тут вдруг на паре по
математике преподаватель просит всех посчитать число Фибоначчи для
100. Кто-то будет считать вручную (так точно не нужно), кто-то
посчитает на калькуляторе, а кто-то подумает, что он самый крутой и
напишет рекурсивную программу на Python и немного огорчится,
потому что данная программа будет достаточно долго считаться, если
ее просто так запускать. Но именно тут к вам на помощь приходят
декораторы, например @lru_cache (он предназначен для решения задач
динамическим программированием, если простыми словами, то этот
декоратор запоминает промежуточные результаты и при рекурсивном
вызове функции программа не будет считать одни и те же значения, а
просто “возьмёт их из этого декоратора”). Вам нужно написать
программу, которая будет считать числа Фибоначчи для 100 и
запустить ее без этого декоратора и с ним, посмотреть на разницу во
времени решения поставленной задачи.
P.S. при запуске без декоратора можете долго не ждать, для
наглядности хватит 10 секунд ожидания

#### Выполнение:
```python
from functools import lru_cache

@lru_cache(None)
def fibonache(n):
    if n==0:
        return 0
    elif n==1:
        return 1
    return fibonache(n-1)+fibonache(n-2)


if __name__=='__main__':
    print(fibonache(100))
```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/e0d28971-aa55-4646-a257-400daed66a5a)

## Задание 2
Илья пишет свой сайт и ему необходимо сделать минимальную
проверку ввода данных пользователя при регистрации. Для этого он
реализовал функцию, которая выводит данные пользователя на экран и
решил, что будет проверять правильность введённых данных при
помощи декоратора, но в этом ему потребовалась ваша помощь.
Напишите декоратор для функции, который будет принимать все
параметры вызываемой функции (имя, возраст) и проверять чтобы
возраст был больше 0 и меньше 130.
Причем заметьте, что неважно сколько пользователь введет данных на
сайт к Илье, будут обрабатываться только первые 2 аргумента
#### Выполнение:
```python
def check(input_func):
    def output_func(*args):
        name, age = args[0], args[1]

        if age<0 or age >130:
            age='Недопустимый возраст'
        input_func(name, age)

    return output_func

@check
def personal_info(name, age):
    print(f"Name {name} Age: {age}")


if __name__ =="__main__":
    personal_info('Владимип', 38)
    personal_info('Александр', -5)
    personal_info('Петр', 138,15,48,2)

```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/53b5cc0e-6ac0-4b88-856b-92513fc24403)


## Задание 3
Вам понравилась идея Ильи с сайтом, и вы решили дальше работать
вместе с ним. Но вот в вашем проекте появилась проблема, кто-то
пытается сломать вашу функцию с получением данных для сайта. Эта
функция работает только с данными integer, а какой-то недохакер
пытается все сломать и вместо нужного типа данных отправляет string.
Воспользуйтесь исключениями, чтобы неподходящий тип данных не
ломал ваш сайт.
Также дополнительно можете обернуть весь код функции в
try/except/finally для того, чтобы программа вас оповестила о том, что
выявлена какая-то ошибка или программа успешно выполнена.
#### Выполнение:
```python
def data(*args):
    try:
        for i in range(len(*args)):
            try:
                result=(args[0][i]*15)//10
                print(result)
            except Exception as ex:
                print(ex)
    except Exception as ex:
        print(ex)
    finally:
                print('Вся инфа обработана')

if __name__=='__main__':
    data([1,15,'Hello','gd','doof','fasdc',23,52])

```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/1638728e-8709-4b75-9a4f-67fdc9deedec)

## Задание 4
Продолжая работу над сайтом, вы решили написать собственное
исключение, которое будет вызываться в случае, если в функцию
проверки имени при регистрации передана строка длиннее десяти
символов, а если имя имеет допустимую длину, то в консоль
выводиться “Успешная регистрация”

#### Выполнение:
```python
class NegativeValueException(Exception):
    pass

def check_name(name):
    if len(name)>10:
        raise NegativeValueException('Длина более 10 символов')
    else:
        print('Успешеая регистрация')

if __name__ == '__main__':
    name='123456788889203'
    check_name(name)

```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/da1764a4-5f09-456d-8d51-37bf8f37c68c)

## Задание 5
После запуска сайта вы поняли, что вам необходимо добавить логгер,
для отслеживания его работы. Готовыми вариантами вы не захотели
пользоваться, и поэтому решили создать очень простую пародию. Для
этого создали две функции: __init__() (вызывается при создании класса
декоратора в программе) и __call__() (вызывается при вызове
декоратора). Создайте необходимый вам декоратор. Выведите все логи
в консоль

#### Выполнение:
```python
class SiteChecker:
    def __init__(self,func):
        print('> Класс SiteChecker метод __init__ успешный запуск')
        self.func = func

    def __call__(self):
        print('> ПРоверка перед запуском', self.func.__name__)
        self.func()
        print('> Проверка безопасного выключения')


@SiteChecker
def site():
    print('Усердная работа сайта')

if __name__ == '__main__':
    print('> Сайт запущен')
    site()
    print('> Сайт выключен')

```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/7d83996d-3043-44b8-a12e-1022d23ea343)


## Самостоятельная работа 
## Задание №1
Вовочка решил заняться спортивным программированием на python, но
для этого он должен знать за какое время выполняется его программа.
Он решил, что для этого ему идеально подойдет декоратор для
функции, который будет выяснять за какое время выполняется та или
иная функция. Помогите Вовочке в его начинаниях и напишите такой
декоратор.
Подсказка: необходимо использовать модуль time
Декоратор необходимо использовать для этой функции

#### Выполнение:
```python
import time

def timing_decorator(func):
    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        execution_time = end_time - start_time
        print(f"Время выполнения функции {func.__name__}: {execution_time} секунд")
        return result
    return wrapper

@timing_decorator
def fibonacci():
    fib1 = fib2 = 1
    for i in range(2, 200):
        fib1, fib2 = fib2, fib1 + fib2
        print(fib2, end=' ')

if __name__ == '__main__':
    fibonacci()

```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/11de273d-b1ea-466e-bf34-4b9708b81975)

### Вывод:  декоратор timing_decorator принимает функцию func и возвращает новую функцию wrapper, которая измеряет время выполнения func. Декоратор затем применяется к функции fibonacci с помощью символа @. Когда fibonacci() вызывается, декоратор автоматически измеряет время выполнения этой функции и выводит результат.

## Задание №2
Посмотрев на Вовочку, вы также загорелись идеей спортивного
программирования, начав тренировки вы узнали, что для решения
некоторых задач необходимо считывать данные из файлов. Но через
некоторое время вы столкнулись с проблемой что файлы бывают
пустыми, и вы не получаете вводные данные для решения задачи.
После этого вы решили не просто считывать данные из файла, а всю
конструкцию оборачивать в исключения, чтобы избежать такой проблемы. Создайте пустой файл и файл, в котором есть какая-то
информация. Напишите код программы. Если файл пустой, то, нужно
вызвать исключение (“бросить исключение”) и вывести в консоль
“файл пустой”, а если он не пустой, то вывести информацию из файла.

#### Выполнение:
```python
def read_file(file_path):
    try:
        with open(file_path, 'r') as file:
            content = file.read()
            if not content:
                raise Exception("Файл пустой")
            else:
                print("Информация из файла:")
                print(content)
    except FileNotFoundError:
        print(f"Файл '{file_path}' не найден")
    except Exception as e:
        print(f"Исключение: {e}")

if __name__ == '__main__':
    # Пример использования для пустого файла
    print("Для пустого файла:")
    read_file('empty_file.txt')

    print("\n" + "="*30 + "\n")

    # Пример использования для файла с информацией
    print("Для файла с информацией:")
    read_file('non_empty_file.txt')


```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/7e9d05cd-6471-4a3c-902e-58f16b33adf5)
![image](https://github.com/MatSs1/program_ingener/assets/106054898/126b4e4f-7c13-4050-b5b6-a90b61951f17)
![image](https://github.com/MatSs1/program_ingener/assets/106054898/085c54cd-bcb1-4c62-b7e1-26f9edac0675)

### Вывод: Этот код определяет функцию read_file, которая сначала пытается открыть файл, читает его содержимое и проверяет, не является ли файл пустым. Если файл пустой, генерируется исключение с сообщением "Файл пустой". Если файл не пустой, выводится содержимое файла. Если файл не найден, обрабатывается исключение FileNotFoundError.


## Задание №3
Напишите функцию, которая будет складывать 2 и введенное
пользователем число, но если пользователь введет строку или другой
неподходящий тип данных, то в консоль выведется ошибка
“Неподходящий тип данных. Ожидалось число.”. Реализовать
функционал программы необходимо через try/except и подобрать
правильный тип исключения. Создавать собственное исключение
нельзя. Проведите несколько тестов, в которых исключение вызывается
и нет. Результатом выполнения задачи будет листинг кода и
получившийся вывод в консоль

#### Выполнение:
```python
def add_two_and_number():
    try:
        user_input = input("Введите число: ")
        result = 2 + float(user_input)
        print(f"Результат сложения: {result}")
    except ValueError:
        print("Неподходящий тип данных. Ожидалось число.")

if __name__ == '__main__':
    # Тест 1: ввод корректного числа
    add_two_and_number(10)

    # Тест 2: ввод строки вместо числа
    add_two_and_number("fjskf")

    # Тест 3: ввод другого неподходящего типа данных
    add_two_and_number("-2")

```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/3deb5816-0216-4cf5-986f-c6e6f2bda373)


### Вывод:используется исключение ValueError, которое возникает, если происходит ошибка преобразования строки в число с помощью float(). Если пользователь вводит корректное число, программа выполняется успешно. Если пользователь вводит строку или другой неподходящий тип данных, перехватывается исключение, и выводится соответствующее сообщение об ошибке.


## Задание №4
Создайте собственный декоратор, который будет использоваться для
двух любых вами придуманных функций. Декораторы, которые
использовались ранее в работе нельзя воссоздавать. Результатом
выполнения задачи будет: класс декоратора, две как-то связанными с
ним функциями, скриншот консоли с выполненной программой и
подробные комментарии, которые будут описывать работу вашего
кода.

#### Выполнение:
```python
  import time

class TimingDecorator:
    def __init__(self, func):
        self.func = func

    def __call__(self, *args, **kwargs):
        start_time = time.time()
        result = self.func(*args, **kwargs)
        end_time = time.time()
        execution_time = end_time - start_time
        print(f"Время выполнения функции {self.func.__name__}: {execution_time} секунд")
        return result

@TimingDecorator
def factorial(n):
    result = 1
    for i in range(1, n + 1):
        result *= i
    return result

@TimingDecorator
def generate_squares(n):
    return [i ** 2 for i in range(1, n + 1)]

if __name__ == '__main__':
    print(factorial(5))

    print(generate_squares(5))

```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/9bd3b3f2-b61a-4df5-88c0-e347c6973bb5)


### Вывод: В этом коде создается класс TimingDecorator, который служит в качестве декоратора. Он принимает функцию в конструкторе и переопределяет метод __call__, чтобы измерять время выполнения переданной функции.
Далее, используем этот декоратор для двух функций: factorial (вычисление факториала) и generate_squares (генерация списка квадратов чисел). Результат выполнения каждой функции выводится в консоль, включая время выполнения, измеренное с использованием нашего декоратора.


## Задание №5
Создайте собственный декоратор, который будет использоваться для
двух любых вами придуманных функций. Декораторы, которые
использовались ранее в работе нельзя воссоздавать. Результатом
выполнения задачи будет: класс декоратора, две как-то связанными с
ним функциями, скриншот консоли с выполненной программой и
подробные комментарии, которые будут описывать работу вашего
кода.

#### Выполнение:
```python
class CustomError(Exception):
    def __init__(self, message="Произошла ошибка"):
        self.message = message
        super().__init__(self.message)

def process_user_data(name, age):
    try:
        if not name or not age:
            raise CustomError("Имя и возраст должны быть заполнены")

        if not isinstance(age, int) or age < 0:
            raise CustomError("Некорректный формат возраста")

        print(f"Данные пользователя: Имя - {name}, Возраст - {age} лет")
    except CustomError as e:
        print(f"Ошибка при обработке данных пользователя: {e}")

def read_file_content(file_path):
    try:
        with open(file_path, 'r') as file:
            content = file.read()
            if not content:
                raise CustomError("Файл пустой")
            else:
                print(f"Содержимое файла: {content}")
    except CustomError as e:
        print(f"Ошибка при чтении файла: {e}")

if __name__ == '__main__':
    # Тест 1: Правильные данные пользователя
    process_user_data("John Doe", 30)

    # Тест 2: Некорректный возраст (вызовет исключение)
    process_user_data("Alice", -25)

    # Тест 3: Пустое имя пользователя (вызовет исключение)
    process_user_data("", 22)

    # Тест 4: Чтение содержимого файла
    read_file_content("example_file.txt")

    # Тест 5: Чтение из пустого файла (вызовет исключение)
    read_file_content("empty_file.txt")

```
#### Результат:
![image](https://github.com/MatSs1/program_ingener/assets/106054898/61cd2c70-4b90-4f53-9cf9-26e2cb594d73)
![image](https://github.com/MatSs1/program_ingener/assets/106054898/445947e6-f4eb-4d50-91e4-2f7d8333f971)

### Вывод: В этом коде используется исключение CustomError для обработки ошибок в двух сценариях: при обработке данных пользователя и при чтении файла. Если возникает ошибка, программа выводит соответствующее сообщение.


## Общий вывод: В Python декораторы и исключения являются важными элементами языка, предоставляющими эффективные и гибкие механизмы для организации кода и обработки ошибок.
